### PR TITLE
Feature/more flexible as provider

### PR DESCRIPTION
**Pull request by l-1squared**
Missing:
[x] documentation
[x] tests
[x] works for method names.
[x] check once for a real JGiven test
[x] double check using As provider for method descriptions -- getDescription uses it. If anything we're more correct now

### CODE REVIEW

File: docs/stages.adoc [hunk 1]
```diff
@@ -36,9 +36,10 @@ I fill out the registration form
 ----
 The nested steps will appear as a collapsible group in the HTML report.
 
+[#_overriding_the_default_reporting]
 === Overriding the Default Reporting
 
-Sometimes it is necessary to override the default way of the step reporting. For example, if you want to use special characters that are not allowed in Java methods names, such as `(`, `+`, or `%`. You can then use the `@As` annotation to specify what JGiven should put into the report.
+Sometimes it is necessary to override the default way of the step reporting.For example, if you want to use special characters that are not allowed in Java methods names, such as `(`, `+`, or `%`.You can then use the `@As` annotation to specify what JGiven should put into the report.
```
Comment (fudler) [inline docs/stages.adoc:42]: ... reporting. [space] For example ....
... such as (, + or % [comma], [you] can ...

File: docs/stages.adoc [hunk 2]
```diff
@@ -66,6 +67,7 @@ public MyStage some_step() {
     return self();
 }
 ----
+In additon, the `@As` annotation allows declaring an `AsProvider` that handles the step name generation. A default provider can be set via in a JGiven configuration.
```
Comment (fudler) [inline docs/stages.adoc:70]: can be set via [???] in a ...

File: jgiven-core/src/main/java/com/tngtech/jgiven/config/AbstractJGivenConfiguration.java [hunk 3]
```diff
@@ -96,4 +101,20 @@ public Formatter<?> getFormatter( final Class<?> typeToBeFormatted ) {
...
-}
\ No newline at end of file
+
+    /**
+     * Set a default interpreter for method names for all items that are subject to this configuration.
+     * Can be overriden by {@link com.tngtech.jgiven.annotation.As} annotation.
+     * @see com.tngtech.jgiven.annotation.As
+     */
+    public void setAsProvider(AsProvider asProvider) {
+        if(asProvider == null) {
```
Comment (fudler) [inline jgiven-core/src/main/java/com/tngtech/jgiven/config/AbstractJGivenConfiguration.java:111]: if [space]

**Review by fudler — COMMENTED** [review]
First comments, not yet finished.

File: docs/stages.adoc [hunk 4]
```diff
@@ -66,6 +67,7 @@ public MyStage some_step() {
     return self();
 }
 ----
+In additon, the `@As` annotation allows declaring an `AsProvider` that handles the step name generation. A default provider can be set via `@JGivenConfiguration` annotation in the class referenced there.
```
Comment (fudler) [inline docs/stages.adoc:70]: In addition, ...

File: jgiven-core/src/test/java/com/tngtech/jgiven/impl/ConfigurationTest.java [hunk 5]
```diff
@@ -1,36 +1,52 @@
...
+    public void testAsProviderConfiguration() throws Throwable{
+        var testModel = new ReportModel();
+        testModel.setTestClass(getClass());
+        getScenario().setModel(testModel);
+        getScenario().startScenario( this.getClass(), ConfigurationTest.class.getMethod( "testAsProviderConfiguration" ), List.of());
+        given().some_step( new FooParam() );
+        given().another_step( new FooParam() );
+
+        given().some_step( new FooParam() );
+        given().another_step( new FooParam() );
```
Comment (fudler) [inline jgiven-core/src/test/java/com/tngtech/jgiven/impl/ConfigurationTest.java:38]: The duplicate is intended?

**Review by fudler — APPROVED** [review]
But check the comments, perhaps another commit is necessary.

File: docs/stages.adoc (hunk 4 above)
Comment (l-1squared) [reply docs/stages.adoc:70]: took real long to find this

**Review by fudler — APPROVED** [review]

**Pull request merged**